source/opt/dominator_tree.cpp

```diff
@@ -0,0 +1,395 @@
+    };
+  }
+
+  // Returns a vector of the list of successor nodes from a given node.
+  GetBlocksFunction GetSuccessorFunctor() {
+    return [&](const BasicBlock* BB) {
+      BasicBlockListTy* v = &Successors[BB];
+      return v;
+    };
+  }
+
+ private:
+  bool InvertGraph;
+  BasicBlockMapTy Successors;
+  BasicBlockMapTy Predecessors;
+
+  // Build a bi-directional graph from the CFG of F.
+  // If InvertGraph is true, all edge are reverted (successors becomes
+  // predecessors and vise versa).
+  // For convenience, the start of the graph is dummyStartNode. The dominator
+  // tree construction requires a unique entry node, which cannot be guarantied
+  // for the postdominator graph. The dummyStartNode BB is here to gather all
+  // entry nodes.
+  void CreateSuccessorMap(Function& F, BasicBlock* dummyStartNode);
+};
+
+template <typename BBType>
+BasicBlockSuccessorHelper<BBType>::BasicBlockSuccessorHelper(
+    Function& func, BasicBlock* dummyStartNode, bool Invert)
+    : InvertGraph(Invert) {
+  CreateSuccessorMap(func, dummyStartNode);
+}
+
+template <typename BBType>
+void BasicBlockSuccessorHelper<BBType>::CreateSuccessorMap(
+    Function& F, BasicBlock* dummyStartNode) {
+  std::map<uint32_t, BasicBlock*> IDtoBBMap;
+  auto GetSuccessorBasicBlock = [&](uint32_t successorID) {
+    BasicBlock*& Succ = IDtoBBMap[successorID];
+    if (!Succ) {
+      for (BasicBlock& BBIt : F) {
+        if (successorID == BBIt.id()) {
+          Succ = &BBIt;
+          break;
+        }
+      }
+    }
+    return Succ;
+  };
+
+  if (InvertGraph) {
+    // For the post dominator tree, we see the inverted graph.
+    // Successors in the inverted graph are the predecessors in the CFG.
+    // The tree construction requires 1 entry point, so we add a dummy node
+    // that is connected to all function exiting basic blocks.
+    // An exiting basic block is a block with an OpKill, OpUnreachable,
+    // OpReturn or OpReturnValue as terminator instruction.
+    for (BasicBlock& BB : F) {
+      if (BB.hasSuccessor()) {
+        BasicBlockListTy& PredList = Predecessors[&BB];
+        BB.ForEachSuccessorLabel([&](const uint32_t successorID) {
+          BasicBlock* Succ = GetSuccessorBasicBlock(successorID);
+          // Inverted graph: our successors in the CFG
+          // are our predecessors in the inverted graph.
+          Successors[Succ].push_back(&BB);
+          PredList.push_back(Succ);
+        });
+      } else {
+        Successors[dummyStartNode].push_back(&BB);
+        Predecessors[&BB].push_back(dummyStartNode);
+      }
+    }
+  } else {
+    // Technically, this is not needed, but it unifies
+    // the handling of dominator and postdom tree later on.
+    Successors[dummyStartNode].push_back(F.entry().get());
+    Predecessors[F.entry().get()].push_back(dummyStartNode);
+    for (BasicBlock& BB : F) {
+      BasicBlockListTy& SuccList = Successors[&BB];
+
+      BB.ForEachSuccessorLabel([&](const uint32_t successorID) {
```

Comment:
This does not take into account "OpSelectionMerge" instructions. It is possible that the merge point is not technically reachable, but may still need to be in the dominator tree. I think this depends on how we are going to use dominator tree. We should have a discussion on how to handle this. I'll try to come up with some odd cases and see what we should be doing about them.